packages/webpack/lib/plugins/stats.js

```diff
@@ -45,48 +47,51 @@ const extractFiles = (chunkData, rawPublicPath) => {
   };
 };
 
-exports.StatsPlugin = class StatsPlugin {
-  constructor(enhancedPromise) {
+exports.StatsWritePlugin = class StatsWritePlugin {
+  constructor(enhancedPromise, statsFile) {
     this.apply = (compiler) => {
-      compiler.hooks.compilation.tap('StatsPlugin', (compilation) => {
-        compilation.hooks.additionalAssets.tap('StatsPlugin', () => {
+      compiler.hooks.compilation.tap('StatsWritePlugin', (compilation) => {
+        compilation.hooks.additionalAssets.tap('StatsWritePlugin', () => {
           const { publicPath } = compilation.outputOptions;
 
           if (compilation.compiler.isChild()) {
             return;
           }
 
-          try {
-            enhancedPromise.resolve({
-              ...compilation
-                .getStats()
-                .toJson({ all: false, assets: true, entrypoints: true }),
-              ...extractFiles(analyzeCompilation(compilation), publicPath),
-            });
-          } catch (error) {
-            enhancedPromise.reject(error);
-          }
+          const stats = {
+            ...compilation
+              .getStats()
+              .toJson({ all: false, assets: true, entrypoints: true }),
+            ...extractFiles(analyzeCompilation(compilation), publicPath),
+          };
+
+          mkdirSync(dirname(statsFile), { recursive: true });
+          writeFileSync(statsFile, JSON.stringify(stats), 'utf8');
+
+          enhancedPromise.resolve(stats);
         });
       });
-
-      compiler.hooks.watchRun.tap('StatsPlugin', () => enhancedPromise.reset());
     };
   }
 };
 
-exports.StatsFilePlugin = class StatsFilePlugin {
-  constructor(enhancedPromise, { statsFile }) {
+exports.StatsAwaitPlugin = class StatsAwaitPlugin {
```

Comment:
**question:** Where is this plugin being used?

Reply:
We still export this (marked as internal API), the name has changed though. But we can also just completely remove it as its implementation is quite simple an we are also providing access to the promise via a mixin hook. So if someone still needs this API he could copy the code.

Reply:
Yes, if it isn't used anywhere we should delete it. Internal API changes are not breaking changes, that's why we've introduced the `.internal` export.

Reply:
Sure, fine for me